tests/rptest/services/action_injector.py

```diff
@@ -0,0 +1,330 @@
+        for entry in action_log:
+            node_final_state[entry.node] = entry.is_reverse_action
+
+        nodes_to_restore = {
+            node
+            for node, is_reversed in node_final_state.items()
+            if not is_reversed
+        }
+
+        hostnames = {node.account.hostname for node in nodes_to_restore}
+        self.redpanda.logger.info(f'Restoring state on {hostnames}')
+        self.do_restore_nodes(nodes_to_restore)
+
+
+class NodeDecommission(DisruptiveAction):
+    def __init__(
+        self,
+        redpanda: RedpandaService,
+        config: ActionConfig,
+        admin: Admin,
+    ):
+        super().__init__(redpanda, config, admin)
+
+    def max_affected_nodes_reached(self):
+        return len(self.affected_nodes) >= self.config.max_affected_nodes
+
+
+class LeadershipTransfer(DisruptiveAction):
+    def __init__(
+        self,
+        redpanda: RedpandaService,
+        config: ActionConfig,
+        admin: Admin,
+        topics: List[TopicSpec],
+    ):
+        super().__init__(redpanda, config, admin)
+        self.topics = topics
+        self.is_reversible = False
+
+    def max_affected_nodes_reached(self):
+        return False
+
+
+class ProcessKill(DisruptiveAction):
+    PROCESS_START_WAIT_SEC = 20
+    PROCESS_START_WAIT_BACKOFF = 2
+
+    def __init__(self, redpanda: RedpandaService, config: ActionConfig,
+                 admin: Admin):
+        super(ProcessKill, self).__init__(redpanda, config, admin)
+        self.failure_injector = FailureInjector(self.redpanda)
+        self.is_reversible = True
+
+    def max_affected_nodes_reached(self):
+        return len(self.affected_nodes) >= self.config.max_affected_nodes
+
+    def do_action(self):
+        node = self.target_node()
+        if node:
+            self.redpanda.logger.info(
+                f'executing action on {node.account.hostname}')
+            self.failure_injector.inject_failure(
+                FailureSpec(FailureSpec.FAILURE_KILL, node))
+            self.affected_nodes.add(node)
+            self.last_affected_node = node
+
+            # Update started_nodes so storage validations are run
+            # on the correct set of nodes later.
+            self.redpanda.remove_from_started_nodes(node)
+            return node
+        else:
+            self.redpanda.logger.warn(f'no usable node')
+
+    def do_reverse_action(self):
+        self._start_rp(node=self.last_affected_node)
+        self.affected_nodes.remove(self.last_affected_node)
+        self.redpanda.add_to_started_nodes(self.last_affected_node)
+
+        last_affected_node, self.last_affected_node = self.last_affected_node, None
+        return last_affected_node
+
+    def do_restore_nodes(self, nodes_to_restore: Set[ClusterNode]):
+        """
+        Attempt to restore the redpanda process on all nodes where it was stopped.
+        """
+        for node in nodes_to_restore:
+            self._start_rp(node)
```

Comment:
this looks error prone, during each action we return an affected node, all the affected nodes are accumulated to action_log which eventually is passed to do_restore_nodes; the problem is that ProcessKill is reversible and all but the last actions are already reverted; so we'll try to reverse it twice;

I recommend to remove the do_restore_nodes, restore_state_on_exit, action_log and all the corresponding config and use just the restore method

Reply:
The `restore_nodes` function makes sure that only the nodes which have not been effectively reversed are restored, please see comment above. The log is processed in `restore_nodes` to find which nodes have already been recovered and skips them.